runtime/compiler/optimizer/VectorAPIExpansion.cpp

```diff
@@ -504,23 +527,60 @@ TR_VectorAPIExpansion::getDataTypeFromClassNode(TR::Compilation *comp, TR::Node
 
          uintptr_t javaLangClass = comp->getKnownObjectTable()->getPointer(symRef->getKnownObjectIndex());
          J9Class *j9class = (J9Class *)(intptr_t)fej9->getInt64Field(javaLangClass, "vmRef");
-         J9JavaVM *vm = fej9->getJ9JITConfig()->javaVM;
-
-         if (j9class == vm->floatReflectClass)
-            return TR::Float;
-         else if (j9class == vm->doubleReflectClass)
-            return TR::Double;
-         else if (j9class == vm->byteReflectClass)
-            return TR::Int8;
-         else if (j9class == vm->shortReflectClass)
-            return TR::Int16;
-         else if (j9class == vm->intReflectClass)
-            return TR::Int32;
-         else if (j9class == vm->longReflectClass)
-            return TR::Int64;
+
+         return j9class;
          }
       }
-   return TR::NoType;
+   return NULL;
+   }
+
+
+TR::DataType
+TR_VectorAPIExpansion::getDataTypeFromClassNode(TR::Compilation *comp, TR::Node *classNode)
+   {
+   J9Class *j9class = getJ9ClassFromClassNode(comp, classNode);
+
+   if (!j9class) return TR::NoType;
+
+   TR_J9VMBase *fej9 = comp->fej9();
+   J9JavaVM *vm = fej9->getJ9JITConfig()->javaVM;
+
+   if (j9class == vm->floatReflectClass)
+      return TR::Float;
+   else if (j9class == vm->doubleReflectClass)
+      return TR::Double;
+   else if (j9class == vm->byteReflectClass)
+      return TR::Int8;
+   else if (j9class == vm->shortReflectClass)
+      return TR::Int16;
+   else if (j9class == vm->intReflectClass)
+      return TR::Int32;
+   else if (j9class == vm->longReflectClass)
+      return TR::Int64;
+   else
+      return TR::NoType;
+   }
+
+
+TR_VectorAPIExpansion::vapiObjType
+TR_VectorAPIExpansion::getObjectTypeFromClassNode(TR::Compilation *comp, TR::Node *classNode)
+   {
+   J9Class *j9class = getJ9ClassFromClassNode(comp, classNode);
+
+   if (!j9class) return Unknown;
+
+   J9UTF8 *className = J9ROMCLASS_CLASSNAME(j9class->romClass);
+   int32_t length = J9UTF8_LENGTH(className);
+   char *classNameChars = (char*)J9UTF8_DATA(className);
+
+   if (!strncmp(classNameChars + length - 6, "Vector", 6))
+      return Vector;
+   else if (!strncmp(classNameChars + length - 4, "Mask", 4))
+      return Mask;
+   else if (!strncmp(classNameChars + length - 7, "Shuffle", 7))
+      return Shuffle;
```

Comment:
is this a potential bug, if the class name is something like MyMask, YourVector, etc etc? since MyMask will be matched with Mask

Reply:
There can be only very limited number of class names for this argument of the intrinsic. But I think the name include a variable part: species name. Let me see if I can make it more clear but avoid too much parsing as well.

Reply:
Added a comment explaining why this comparison is sufficient.